### PR TITLE
Pull Req: Minification of JS & CSS content

### PR DESCRIPTION
**Pull request by adrianlee**
- [Issue #116](https://github.com/blackberry/BB10-Webworks-Packager/issues/116)
- [CI build](http://ci0000003863287:8080/hudson/job/BB10-Webworks-Packager-next-minify/)
- [Docs Issue](http://github.rim.net/Microsites/HTML5-WebWorks/issues/194#pull_comment_form)

### CODE REVIEW

**Comment by adrianlee** [comment]
Got to fix something... Please don't test this yet.

File: Framework [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit 6e3ee97b436cd6d47a37b8378df87d1d72de1b88
+Subproject commit 11c4506d1c5f7ee52ecccf09f3cfd8514509d7de
```
Comment (jeffheifetz) [inline Framework]: WHy are you updating the framework hash?

File: package.json [hunk 2]
```diff
@@ -21,7 +21,10 @@
     "xml2js": "0.1.13",
     "fsext": "0.0.4",
     "validator": "0.4.1",
-    "jWorkflow": "0.6.0"
+    "jWorkflow": "0.6.0",
+    "uglify-js": "1.3.2",
+    "clean-css": "0.4.2",
+    "async": "0.1.22"
```
Comment (jeffheifetz) [inline package.json:27]: These are OSS concerns because they are deployed with our tool, correct @nukulb

**Comment by jeffheifetz** [comment]
You need to update the configure script with the new npm packages. This will require testing on both Mac and Windows.

**Comment by nukulb** [comment]
the configure script does a npm install which pulls from package.json

**Comment by adrianlee** [comment]
@rwmtse @issp 

This is ready for code review and testing. Docs will come shortly

File: test/unit/lib/session.js [hunk 3]
```diff
@@ -95,4 +95,33 @@ describe("Session", function () {
...
         expect(result.outputDir).toEqual(path.normalize(path.join(bbwpDir, "myOutput")));
     });
+
+    it("sets minification by default", function () {
+        var data = {
+            args: [ 'C:/sampleApp/sample.zip' ]
+        },
+        result = session.initialize(data);
+
+        expect(result.noMinify).toEqual(true);
```
Comment (rwmtse) [inline test/unit/lib/session.js]: If cmdline.minify defaults to true, then we should have session.minify defaults to true... this logic is easier to follow

**Comment by rwmtse** [comment]
You need to rebase

**Comment by rwmtse** [comment]
Code looks good to me

**Comment by adrianlee** [comment]
Updated a line to cover edge cases

**lib/session.js:**

``` javascript
"minify": (cmdline.minify === undefined ? true : !!cmdline.minify) && !cmdline.debug,
```

**Comment by rwmtse** [comment]
Please rebase

**Comment by adrianlee** [comment]
rebased. 

@issp can you continue your testing with [this build](http://ci0000003863287:8080/hudson/job/BB10-Webworks-Packager-next-minify/851/)

**Comment by nukulb** [comment]
@adrianlee -can't fold this till you do a performance test

**Comment by nukulb** [comment]
@adrianlee ?? How goes the performance testing?

**Comment by nukulb** [comment]
@kwallis - there seems to be no performance gain because of this ? do we really care for this then?

**Comment by kwallis** [comment]
## What about bar file size? I assume since it is basically a zip, there is no size gain either?

Ken Wallis

Product Manager – BlackBerry WebWorks

Research In Motion

(905) 629-4746 x14369

**Comment by adrianlee** [comment]
With minify, we shed about 100KB off our framework content of a bar.

**Comment by nukulb** [comment]
didn't realize that. That sounds not bad.

**Comment by kwallis** [comment]
## Seems like a win.

Ken Wallis

Product Manager – BlackBerry WebWorks

Research In Motion

(905) 629-4746 x14369

**Comment by issp** [comment]
@nukulb Do I need to retest this feature branch?

**Comment by nukulb** [comment]
No
~Nukul

**Comment by nukulb** [comment]
have'n't made up our mind on benefit.

**Comment by rcruz** [comment]
Rebased off next-zygote which will go into next today pending testing.  Putting this in will save at least 0.5 seconds on boot time.

**Comment by rcruz** [comment]
Introduces a new module called "async"

File: lib/bbwp.js [hunk 4]
```diff
@@ -31,7 +31,7 @@ try {
     cmdline.parse(process.argv);
```
Comment (jeffheifetz) [inline lib/bbwp.js]: Please revert this file

Comment (rcruz) [inline lib/bbwp.js]: Reverted

**Comment by nukulb** [comment]
@rcruz can you get this tested again please and then we will pull it in?
you will need to write a micro site issue for it so that the docs can be updated.

**Comment by nukulb** [comment]
perhaps another way is to minify the framework at build time and not complicate the packager at all? we will get the benefits and not mess with users content.
@rcruz @jeffheifetz @erikj54 thoughts?

**Comment by ejzn** [comment]
We could. Might be nice to have a un-minified version for debugging. If we can do both at build time then I'm game. Jake/jake release?

**Comment by nukulb** [comment]
question is what is better, package time vs. build time.

**Comment by cdelcol** [comment]
+1 for build time. I still don't like the idea of ever changing the developer content.

**Comment by kwallis** [comment]
I am ok with this.  Developer can always do minimization of their own content if they choose.

## 

Ken Wallis

Product Manager – BlackBerry WebWorks

Research In Motion

(905) 629-4746 x14369

**Comment by jeffheifetz** [comment]
I personally like package time because its easier for a developer to take advantage of it

**Comment by rcruz** [comment]
I prefer package time because framework/webplatform can still be debugged with -d.  I understand not messing with the developer's code is key as well, since maybe he wants to debug on a release build.  How about the following behaviour:  user content always unmodified regardless of -d, webworks related code minified if packaged without -d, introduce new flag like -m to minify everything including user content.  What do you think?

**Comment by ejzn** [comment]
I like that option the best. They explicitly ask for the minification.

**Comment by nukulb** [comment]
so if this is really needed for performance improvement we should always be minified except when someone is debugging (-d) or they say --no-minify

I prefer this over the explicit minify since this gives a performance boost. I would like to always have this and only turn it off if I am debugging or don't want to minify explicitly.

@kwallis - thoughts?

**Comment by kwallis** [comment]
How much of the performance gain is there from the user content versus the framework code?

## 

Ken Wallis

Product Manager – BlackBerry WebWorks

Research In Motion

(905) 629-4746 x14369

**Comment by rcruz** [comment]
We are waiting for "DocumentLoadFinished" to display the content webview, so the content page has to finish loading before we display it.  The performance gain of minifying the user content depends on how the user has set up their app.

**Comment by kwallis** [comment]
I heard that perhaps this mechanism is changing based on some work from @nukulb from last night?

## 

Ken Wallis

Product Manager – BlackBerry WebWorks

Research In Motion

(905) 629-4746 x14369

**Comment by nukulb** [comment]
Looks like there is no benefit to this, closing this for now. Reopen if we decide to do it.

**Pull request closed without merging**